compiler/rustc_parse/src/parser/path.rs

```diff
@@ -196,13 +196,14 @@ impl<'a> Parser<'a> {
 
         maybe_whole!(self, NtPath, |path| reject_generics_if_mod_style(self, path.into_inner()));
 
-        if let token::Interpolated(nt) = &self.token.kind {
-            if let token::NtTy(ty) = &**nt {
-                if let ast::TyKind::Path(None, path) = &ty.kind {
-                    let path = path.clone();
-                    self.bump();
-                    return Ok(reject_generics_if_mod_style(self, path));
-                }
+        if let Some(MetaVarKind::Ty) = self.token.is_metavar_seq() {
+            let mut snapshot = self.create_snapshot_for_diagnostic();
+            let ty = snapshot
+                .eat_metavar_seq(MetaVarKind::Ty, |this| this.parse_ty_no_question_mark_recover())
```

Comment:
We should immediately reparse the tokens as a path here, without going through a type.
Because that's what we should eventually do everywhere (reparsing tokens ignoring the metavar kind used for previous parsing) and because snapshot restoring is not supposed to happen on a good path.

Reply:
Ok. I've added a new commit that does this. It required adding an `is_path` field to `MetaVarKind::Ty`, because we don't want to try and reparse unless we're certain that it'll succeed.